### DISCUSSION TITLE
React Native Line Parsing

### DISCUSSION THREAD

**Discussion by sgriff96** [remarkjs/remark#1183] [Q&A]
Hi! I am attempting to use this library for React Native (though fully aware that that is not supported so no worries if this isn't possible).

I'm trying to define my own components for each type of markdown, but no matter what I seem to get an error in React Native for trying to render something without a `<Text />` (React Native specific) element.

This is the markdown I'm passing in:
```
# test

test body

[link](https://www.google.com)
```

I've set a component for basically _every_ single option in components at this rate. It seems like it breaks once there is a line break. 

Any single line from here seems to work, ie
`# test`
`test body`
`[link](https://www.google.com)`

But once I make a new line it seems to try to render that without a React Native `<Text>` element.

I do have `br` replaced (`br: ({ children }) => <Text>{children}</Text>,`) but not sure if that's the proper element for this. Can anyone suggest what I might need to do to handle these line breaks?

**Comment by ChristianMurphy** [comment]
Welcome @sgriff96! 👋 

When you say "this library", which are you referring to?
* https://github.com/remarkjs/react-markdown
* https://github.com/remarkjs/react-remark
* https://github.com/rehypejs/rehype-react
* https://github.com/mdx-js/mdx
* or something else entirely?

https://unifiedjs.com/learn/

> Can anyone suggest what I might need to do to handle these line breaks?

Targeting line breaks feels like the wrong way of going about this.
Going back to the original ask:

> I'm trying to define my own components for each type of markdown, but no matter what I seem to get an error in React Native for trying to render text without a Text element.

If you want to wrap text elements, you probably want a `remark` or `rehype` plugin (see: https://unifiedjs.com/learn/).
`text` does exist as a node type in the Abstract Syntax Tree, and you could wrap it in tags which usually wouldn't exist in HTML from that level in content transformations.

**Comment by sgriff96** [comment]
Oh shoot sorry, not super familiar with Github Discussions. I'm referring to [remarkjs/react-markdown](https://github.com/remarkjs/react-markdown). Let me take a look into those plugins again (looked at them briefly from the `react-markdown` README) and get back to you. Thanks for the quick reply!

**Comment by sgriff96** [comment]
Hey @ChristianMurphy could you elaborate a bit on why targeting line breaks seems like the wrong way to do this? The problem seems to be how react-markdown interprets line breaks in the markdown, and is for some reason returning those as something (I don't know what) that's not part of the `components` options and thus rendering somehow what React Native interprets as text but not wrapped in `<Text>` which you must do in RN.

**Reply by ChristianMurphy** [reply]
Breaks are self closing and have no children https://developer.mozilla.org/en-US/docs/Web/HTML/Element/br

**Reply by ChristianMurphy** [reply]
> The problem seems to be how react-markdown interprets line breaks in the markdown, and is for some reason returning those as something (I don't know what) that's not part of the components options

Text is not a component, so it is not part of the component options

**Reply by ChristianMurphy** [reply]
Again please read my initial comment

> If you want to wrap text elements, you probably want a remark or rehype plugin (see: https://unifiedjs.com/learn/).
> `text` does exist as a node type in the Abstract Syntax Tree, and you could wrap it in tags which usually wouldn't exist in HTML from that level in content transformations.

**Reply by sgriff96** [reply]
Sorry I think I didn't explain myself well initially (or I'm not understanding well, reading through some of the content in unifiedjs.com/learn was going over my head). So is `react-markdown` returning a `<br />` for new lines in the markdown then? Unfortunately the error reporting in React Native leaves something to be desired, so I'm not sure exactly what it's reading and erroring on. 

I'm not following what you mean by "Text is not a component, so it is not part of the component options". I know its not part of the `components` options in react-markdown. I was referring to how React Native requires any string to be wrapped in a `<Text />` component (a RN component), so I'm trying to do this for example: `p: ({ children }) => <Text>{children}</Text>,`. Every time I've referred to Text/text etc has been in that context, not as like an html element or markdown element or anything else.

**Reply by ChristianMurphy** [reply]
> react-markdown returning a <br /> for new lines in the markdown then?

This isn't specifical to `react-markdown`, it is how HTML works.

> I was referring to how React Native requires any string to be wrapped in a <Text /> component (a RN component)

I'm well aware.
You need to translate HTML where text is not a component, to react native where it is.
To do so you need to change the structure of the document, which is what the Abstract Syntax Tree can do.

**Comment by wooorm** [comment]
This seems very much like https://github.com/remarkjs/react-markdown/issues/741.

**Comment by mwaleed-39** [comment]
I faced some issue in react native. when we pressed upload image, the error message shows:
CODE:

import React, { useState } from 'react'
import { Alert, Button, View,StyleSheet } from 'react-native'
import * as ImagePicker from 'react-native-image-picker'
import {launchImageLibrary} from 'react-native-image-picker'

const options ={
title: 'Select Image',
type: 'library',
options: {
maxHeight:200,
maxWidth:200,
selectionLimit: 1,
mediaType: 'photo',
includeBase64: false,
}
}

const Request = () => {
const opengallery = async()=>{
const image = await ImagePicker.launchImageLibrary(options);
console.log(image)
}

return (



)
}
export default Request;

const styles = StyleSheet.create({
container:{
flex:1,
alignItems:'center',
justifyContent:'center'
}
})

when we pressed upload button, the error message shows:

Possible Unhandled Promise Rejection (id: 1):
TypeError: Cannot read property 'launchImageLibrary' of null
TypeError: Cannot read property 'launchImageLibrary' of null

I create API:
app.post('/finerequest', (req, res, next) => {

const file = req.files.image;
cloudinary.uploader.upload(file.tempFilePath, (err, result) => {
console.log(result)
const requestId = req.body.request_id
const fine = new Fine({
_id: new mongoose.Types.ObjectId,
request_id: requestId,
image: result.url
})
fine.save()
.then(result => {
res.status(200).json({
message: "Fine and Vehicle Repair Request added",
response: result
})
})
.catch(err => {
res.status(500).json({
message: "Fine and Vehicle request not added",
error: err
})
})
})
})
How to call this API, when we upload image in front end side, the image will be stored in mongodb

**Comment by mwaleed-39** [comment]
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, View } from 'react-native';
import React from 'react';
import Reg_form from './Frontend/Components/RegForm';
import Login_form from './Frontend/Components/Signin';
import ForgotPassword from './Frontend/Components/ForgotPassword';
import { ApplicationProvider, Layout, Text } from '@ui-kitten/components';
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import {
createDrawerNavigator,
DrawerContentScrollView,
DrawerItemList,
DrawerItem,
} from '@react-navigation/drawer';
import * as eva from '@eva-design/eva';
import ChangePassword from './Frontend/Components/ChangePassword';
import VerifyOTP from './Frontend/Components/VerifyOTP';
import Home from './Frontend/Components/Home Screen';
import Profile from './Frontend/Components/Profile';
import Icon from 'react-native-vector-icons'
const Stack = createNativeStackNavigator();
const Drawer = createDrawerNavigator();

const CustomDrawer = props => {

return (
<View style={{ flex: 1 }}>
<DrawerContentScrollView {...props}>
<View
style={{
alignSelf: "center",
padding: 20,

      }}
    >
      <View>
        <Text style={{ fontSize: 18, fontWeight: "bold" }}> App</Text>
      </View>

    </View>
    <DrawerItemList {...props} />
    <LogoutButton iconName="sign-out" color="black" />
  </DrawerContentScrollView>

</View>
);
};

function DrawerRoutes() {

return (

<Drawer.Navigator initialRouteName="Home"

  drawerContent={props => <CustomDrawer {...props} />}
>

  <Drawer.Screen name="Home" component={Home}

    options={{
      headerShown: true,
      
      title: 'Home',
      drawerIcon: ({ focused, size }) => (
        <Icon
          name="home"
          size={size}
          color='black'

        />

      )
    }}
  />
  <Drawer.Screen name="Profile" component={Profile}

    options={{
      headerShown: true,
      title: 'Profile',
      drawerIcon: ({ focused, size }) => (
        <Icon
          name="user"
          size={size}
          color='black'

        />

      )
    }}
  />




</Drawer.Navigator>
);
}

export default function App() {
return (
<ApplicationProvider {...eva} theme={eva.light}>

<NavigationContainer>
<Stack.Navigator>
          <Stack.Screen name="Registration" component={Reg_form} />
          <Stack.Screen name="Log In" component={Login_form}/>
          <Stack.Screen name="Forgot" component={ForgotPassword}/>
          <Stack.Screen name="Password" component={ChangePassword}/>
          <Stack.Screen name="Verification" component={VerifyOTP}/>
          <Stack.Screen name="Home" component={Home}  />
          <Stack.Screen name="DrawerRoutes" component={DrawerRoutes}/>
</Stack.Navigator>

); }
When we log in to the App, then navigate into the Home Screen. Then we have not visible Drawer navigation

**Reply by wooorm** [reply]
What?! Please take time formulating and fixing your q

**Comment by mwaleed-39** [comment]
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, View } from 'react-native';
import React from 'react';
import Reg_form from './Frontend/Components/RegForm';
import Login_form from './Frontend/Components/Signin';
import ForgotPassword from './Frontend/Components/ForgotPassword';
import { ApplicationProvider, Layout, Text } from '@ui-kitten/components';
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import {
createDrawerNavigator,
DrawerContentScrollView,
DrawerItemList,
DrawerItem,
} from '@react-navigation/drawer';
import * as eva from '@eva-design/eva';
import ChangePassword from './Frontend/Components/ChangePassword';
import VerifyOTP from './Frontend/Components/VerifyOTP';
import Home from './Frontend/Components/Home Screen';
import Profile from './Frontend/Components/Profile';
import Icon from 'react-native-vector-icons'
const Stack = createNativeStackNavigator();
const Drawer = createDrawerNavigator();

const CustomDrawer = props => {

return (
<View style={{ flex: 1 }}>
<DrawerContentScrollView {...props}>
<View
style={{
alignSelf: "center",
padding: 20,

      }}
    >
      <View>
        <Text style={{ fontSize: 18, fontWeight: "bold" }}> App</Text>
      </View>

    </View>
    <DrawerItemList {...props} />
    <LogoutButton iconName="sign-out" color="black" />
  </DrawerContentScrollView>

</View>
);
};

function DrawerRoutes() {

return (

<Drawer.Navigator initialRouteName="Home"

  drawerContent={props => <CustomDrawer {...props} />}
>

  <Drawer.Screen name="Home" component={Home}

    options={{
      headerShown: true,
      
      title: 'Home',
      drawerIcon: ({ focused, size }) => (
        <Icon
          name="home"
          size={size}
          color='black'

        />

      )
    }}
  />
  <Drawer.Screen name="Profile" component={Profile}

    options={{
      headerShown: true,
      title: 'Profile',
      drawerIcon: ({ focused, size }) => (
        <Icon
          name="user"
          size={size}
          color='black'

        />

      )
    }}
  />




</Drawer.Navigator>
);
}

export default function App() {
return (
<ApplicationProvider {...eva} theme={eva.light}>

<NavigationContainer>
<Stack.Navigator>
          <Stack.Screen name="Registration" component={Reg_form} />
          <Stack.Screen name="Log In" component={Login_form}/>
          <Stack.Screen name="Forgot" component={ForgotPassword}/>
          <Stack.Screen name="Password" component={ChangePassword}/>
          <Stack.Screen name="Verification" component={VerifyOTP}/>
          <Stack.Screen name="Home" component={Home}  />
          <Stack.Screen name="DrawerRoutes" component={DrawerRoutes}/>
</Stack.Navigator>

); }
When we log into the App, then navigate into the Home Screen. So we have not visible Drawer navigation. What's the problem

**Reply by ChristianMurphy** [reply]
@mwaleed-39 This is the discussion and Q&A forum for `remark` and `react-markdown`.
You don't appear to be using either in your application, from the limited code you have shown.

If you are looking for general React Native support, please see their support channels https://reactnative.dev/community/support

If you are using `react-markdown` or `remark` and believe that it is causing the issue.
Share how you are using these libraries, how you have configured them, and _why_ you think these are the error and not something else in the application.
Following the support guidelines https://github.com/remarkjs/.github/blob/main/support.md

If you continue to spam incomprehensible and unrelated comments at the `remark` community, you will be blocked.
